### PR TITLE
[MFEM] Add sequential version

### PR DESCRIPTION
**Pull request by Antoinemarteau**
Sequential version of MFEM, a Finite Element Method library under BSD license ;).

I did not manage to make it work with expand_cxxstring_abis() that the audit wanted me to add.

https://mfem.org/

### CODE REVIEW

File: M/MFEM/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+sources = [
+    GitSource("https://github.com/mfem/mfem.git", "a1f6902ed72552f3e680d1489f1aa6ade2e0d3b2")
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir
+cd mfem*
+mkdir build; cd build
+cmake -DCMAKE_INSTALL_PREFIX=${prefix}       -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN}       -DCMAKE_BUILD_TYPE=Release       -DBUILD_SHARED_LIBS=1       ..
```
Comment (giordano) [inline M/MFEM/build_tarballs.jl:18]:
```suggestion
cd $WORKSPACE/srcdir/mfem*
mkdir build; cd build
cmake -DCMAKE_INSTALL_PREFIX=${prefix} \
    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
    -DCMAKE_BUILD_TYPE=Release \
    -DBUILD_SHARED_LIBS=1 \
    ..
```

File: M/MFEM/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    Platform("armv6l", "linux"; call_abi = "eabihf", libc = "glibc"),
+    Platform("armv7l", "linux"; call_abi = "eabihf", libc = "glibc"),
+    Platform("powerpc64le", "linux"; libc = "glibc"),
+    Platform("x86_64", "macos"; ),
+    Platform("aarch64", "macos"; ),
+    Platform("x86_64", "freebsd"; ),
+    Platform("i686", "windows"; ),
+    Platform("x86_64", "windows"; )
+]
+
```
Comment (giordano) [inline M/MFEM/build_tarballs.jl:38]: > I did not manage to make it work with expand_cxxstring_abis() that the audit wanted me to add.

I'm not sure what was your problem, but this should be here:
```suggestion
platforms = expand_cxxstring_abis(platforms)
```

File: M/MFEM/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    Platform("aarch64", "linux"; libc = "glibc"),
+    Platform("armv6l", "linux"; call_abi = "eabihf", libc = "glibc"),
+    Platform("armv7l", "linux"; call_abi = "eabihf", libc = "glibc"),
+    Platform("powerpc64le", "linux"; libc = "glibc"),
+    Platform("x86_64", "macos"; ),
+    Platform("aarch64", "macos"; ),
+    Platform("x86_64", "freebsd"; ),
+    Platform("i686", "windows"; ),
+    Platform("x86_64", "windows"; )
+]
```
Comment (giordano) [inline M/MFEM/build_tarballs.jl:37]: Maybe we can build for all platforms and see what's the problem with the failing ones?
```suggestion
platforms = supported_platforms()
```

**Review by giordano — COMMENTED** [review]
Thanks for your contribution!  Some comments below

File: M/MFEM/build_tarballs.jl (hunk 2 above)
Comment (Antoinemarteau) [reply M/MFEM/build_tarballs.jl:38]: MFEM Doc says:
Starting with version 4.0, MFEM requires a C++11 compiler. We recommend using
  a newer compiler, e.g. GCC version 4.9 or higher.

When I tested with expand_cxxstring_abis, all compilations with cxx03 failed.

**Comment by Antoinemarteau** [comment]
> Thanks for your contribution! Some comments below

Thanks :)

The build systematically fail with musl, I don't know what to do…

File: M/MFEM/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# Note that this script can accept some limited command-line arguments, run
+# `julia build_tarballs.jl --help` to see a usage message.
+using BinaryBuilder, Pkg
+
+name = "MFEM_seq"
```
Comment (giordano) [inline M/MFEM/build_tarballs.jl:5]: Note: this name must match exactly the name of the directory, so either change this name or move the directory.

File: M/MFEM/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+products = [
+    LibraryProduct("libmfem", :libmfem)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = Dependency[
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6", preferred_gcc_version=v"5")
```
Comment (giordano) [inline M/MFEM/build_tarballs.jl:47]: Looks like GCC 5 doesn't like boolean enums, but GCC 6 does:
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6", preferred_gcc_version=v"6")
```

```
[10:40:12] /workspace/srcdir/mfem/fem/restriction.cpp: In member function ‘void mfem::NCL2FaceRestriction::ComputeScatterIndicesAndOffsets(mfem::ElementDofOrdering, mfem::FaceType)’:
[10:40:12] /workspace/srcdir/mfem/fem/restriction.cpp:2108:66: error: could not convert ‘(mfem::FaceType)type’ from ‘mfem::FaceType’ to ‘bool’
[10:40:12]                 (type==FaceType::Interior? "interior" : "boundary") <<
[10:40:12]                                                                   ^
```

**Review by giordano — COMMENTED** [review]
I pushed a patch (backport of https://github.com/mfem/mfem/pull/2995) for building with Musl systems

File: M/MFEM/build_tarballs.jl (hunk 4 above)
Comment (Antoinemarteau) [reply M/MFEM/build_tarballs.jl:5]: The idea was to compile MPI MFEM later (it's harder)
Can I use the same structure as MUMPS, that is move this build_tarballs.jl in an inner folder like
M / MFEM / MFEM_seq / build_tarballs.jl
?

Comment (giordano) [reply M/MFEM/build_tarballs.jl:5]: That's OK, yes

File: M/MFEM/MFEM_seq/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+products = [
+    LibraryProduct("libmfem", :libmfem)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = Dependency[
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6", preferred_gcc_version=v"6")
```
Comment (giordano) [inline M/MFEM/MFEM_seq/build_tarballs.jl:47]: GCC 6 runs into an internal compiler error when targeting aarch64-linux-gnu-cxx11. That's not good, we'll have to use a different version again
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6", preferred_gcc_version=v"7")
```

**Comment by Antoinemarteau** [comment]
Alléluia ! Thank you very much @giordano ,  I would not have fixed this alone !

**Comment by giordano** [comment]
The internal compiler errors are very unfortunate but we can do nothing apart from trying a different version: it's a bug in the compiler itself.  But at least now all is working!  Thanks again for your contribution

**Pull request merged**